posts/2022/08-25-adapter_pattern.md

```diff
@@ -0,0 +1,252 @@
+A GenServer is more reliable since we can utilize `Process.send_after/4` to have the GenServer call itself reliably every n milliseconds.
+
+Since the GenServer is doing the same things for both APIs, we can start the same GenServer twice; all we have to do is register the two instances of our GenServer under different names.
+
+We also need code for calling our APIs. I decided to create one API module per queried API. While the two API modules have the same job (calling the API, fetching JSON data, decoding JSON data, standardizing data, returning standardized to the GenServer), they are operating under different circumstances: different HTTP addresses, plus the fact that API B also takes in the additional search parameter. 
+
+This is the big moment: enter Behaviour.
+
+Since we assume that both our API modules will perform the same actions and send data in the same unified format back to the GenServer, we can go one step further and not just assume but require that they do so. 
+To this end, we can use another module to define the desired Behaviour and implement said Behaviour in our API modules. Now, we cannot only be sure that our existing API modules act in a certain way, but we can also easily add further API modules in the future. All we have to do is implement the same Behaviour for them, thus ensuring that we don't accidentally add a rogue API module that acts different from the rest and risks bringing down its associated GenServer.
+
+But enough with the talk. Let's look at some code!
+
+
+# 4. The code
+
+
+**General organization**
+
+```
+
+lib
+│   ├── my_app
+│   │   ├── application.ex
+│   │   ├── config.ex
+│   │   ├── match_data_processors
+│   │   │   ├── apis
+│   │   │   │   ├── client.ex
+│   │   │   │   ├── api_a.ex
+│   │   │   │   ├── api_b.ex
+│   │   │   │   
+│   │   │   └── helpers
+│   │   │       └── processing_helpers.ex
+│   │   ├── matches
+│   │   │   └── match.ex
+│   │   ├── matches.ex
+│   │   ├── match_handler.ex
+
+```
+
+The `match_handler.ex` file is the GenServer module, `match.ex` and `matches.ex` define our the match `Ecto.Schema` and the corresponding context module, and we can also see the `client.ex` (this is where the Behaviour is defined), our two API modules and a `processing_helpers.ex` file, which contains some helper functions accessed by both API modules.
+
+
+**GenServer**
+
+First stop: `application.ex`, where the GenServer is added to the Supervision tree:
+
+```
+defmodule MyApplication.Application do
+  
+  use Application
+
+  @impl true
+  def start(_type, _args) do
+    children = [MyApp.Repo] ++ match_handlers()
+
+    
+    opts = [strategy: :one_for_one, name: MyApp.Supervisor]
+    Supervisor.start_link(children, opts)
+  end
+
+  if Mix.env() === :test do
+    def match_handlers do
+      []
+    end
+  else
+    @api_modules MyApp.Config.api_modules()
```

Comment:
You don't have the `MyApp.Config` module defined in this blog post right?

Reply:
Right. I will hope it is still clear